### DISCUSSION TITLE
How can I create multiple documents from one set of variables?

### DISCUSSION THREAD

**Discussion by stueja** [quarto-dev/quarto-cli#10234] [Q&A]
### Description

In our training for medical IT, we use ~30 workbooks which basically describe the same exercises in slightly different order. The exercises have different parameters per system, e. g. IP address, hostname, etc. Some of these parameters change on a yearly basis, e. g. a certain password.

My predecessor created ~30 single, separate MS Word documents for system1, system2, ... system30 with tables, tables-in-tables, and images in tables and out of tables and in between tables and I have now inherited that mess. I dearly want to convert that reproducibility hell into something which is more adjustable, handle-able and reproducible.

So far, I have tried
1. _a Book project with project profiles._ The `workbook-system1.yml` allows me to define the system's variables as well as the chapter structure. In this case, I would need ~30 project profiles, one per medical system.
2. _a single `.qmd` file with `includes`._ A `workbook-system1.qmd`, just like the Book, allows me to define the system's variables as well as the chapter structure, plus I'd end up with one `.pdf` per `.qmd`. In this case, I'd need ~30 `.qmd` files.
3. _a single `.yml` file with system settings for each system,_ e. g. IP address, plus one `.qmd` file and/or includes which use `{{< var/meta system1.ipaddress >}}` and `{{< var/meta system2.ipaddress >}}` and so on. Here, I'd need one `.yml` file and ~30 `.qmd` files.
4. _The `--metadata` command line argument._ Allows me to be independent of one or many `.yml` files, but takes ~30 calls with ~20 parameters each time. 

I have read https://github.com/quarto-dev/quarto-cli/discussions/7805, https://github.com/quarto-dev/quarto-cli/discussions/7814, https://github.com/quarto-dev/quarto-cli/discussions/2021, but they seem unrelated to my case.

One thing which came to my mind is whether it is possible to loop over the variables of the `.yml` file and create one book project from one iteration. Or whether it is possible to tell quarto to render a book project only with a subset of variables from the `.yml`.

I guess the overarching question is: which is the best way to have the least amount of `.yml` files with the least amount of `.qmd` files?

Thank you.

**Comment by mcanouil** [comment]
I believe the following demo is of interest to you:
- https://github.com/mcanouil/demo-generate-quarto-invoices
It uses one single Quarto document.

You don't necessarily have to use YAML file as you could set individual value from the command line.

**Reply by mcanouil** [reply]
> Or whether it is possible to tell quarto to render a book project only with a subset of variables from the .yml.

Take a look at conditional content and project profiles:
- https://quarto.org/docs/authoring/conditional.html
- https://quarto.org/docs/projects/profiles.html

**Reply by stueja** [reply]
> Take a look at conditional content and project profiles:

I believe this would require documents like

```
:::{.content-visible when-project="system1"}
some text/include for system1
:::

:::{.content-visible when-project="system2"}
some text/include for system2
:::
... and so on for every system...
```
Wouldn't it? And I believe I need that syntax for making the workbooks differ between an expert, intermediate and beginner level, but that is a future step in this project.

> I believe the following demo is of interest to you:
> 
>     * https://github.com/mcanouil/generate-quarto-invoices

This looks very promising. If I understand correctly so far, a kind of pre-processing script before the actual quarto processing takes place. I will have a closer look at that.

The `--metadata-file` switch is nowhere to be found on the quarto website nor in `quarto help render`. Thanks for that, I didn't know it existed.

**Reply by mcanouil** [reply]
> > Take a look at conditional content and project profiles:
> 
> 
> 
> I believe this would require documents like
> 
> 
> 
> ```
> 
> :::{.content-visible when-project="system1"}
> 
> some text/include for system1
> 
> :::
> 
> 
> 
> :::{.content-visible when-project="system2"}
> 
> some text/include for system2
> 
> :::
> 
> ... and so on for every system...
> 
> ```
> 

Don't make up things, simply use what is documented in the page I linked, *i.e.*, `when-profile`: https://quarto.org/docs/projects/profiles.html


The repository I shared is only one of the various ways to achieve what you want.
There is not really one above the other.

**Comment by harrelfe** [comment]
Sometimes looping at the command line level is simplest, e.g.

```
for(….) {
  system(paste(‘quarto render’, …))
}
```
Wwhere `paste(…)` puts together the right `yaml` parameter to hand to the template and specifies the name of the output file for the current loop.
